### PR TITLE
Find optional meeting query

### PR DESCRIPTION
**Pull request by gannerr**
We extend our algorithm to find meetings for optional participants as well.
- Algo code changes:
    - We split the code into separate functions now, to represent each step of the algorithmic process...
    - sortEvents() we find events with the participants we want, and sort them based on start time. O(nlogn) for sorting.
    - mergeEvents() merges overlapping events. The strategy here is to take 2 adjacent events in our list, check if the second one's start time is >= the first one's end time, and if so, merge. We can do this thanks to sortEvents() for sorting events based on start time. 
                      - O(n) total time - O(n) for iterating through list of events, merge takes O(1) time.
    - findAvailableTimeRanges() finds the gaps between events and asserts whether there is enough time for a meeting. The strategy here is to take 2 adjacent events in our list and checks the difference in start time for second one and end time for first one. We also make sure to check for time from beginning of day to start time of first event of the day, as well as time from end of last event of the day to end of day. So long as these difference in times >= meeting time, we found a good meeting time. 
                      - O(n) total time - O(n) for iterating through list of events, comparing time differences with meeting time length takes O(1) time.
    - buildQuery() calls these 3 functions to build our query of good meeting times. 
                         - Overall, O(nlogn) time due to sortEvents().
     - To account for optional participants, we just treat them as mandatory participants and run buildQuery() on them. We check if there is any commonality between the queries for optional participants and actual mandatory participants - if there is, we return those times. Else, we prioritize the actual mandatory participants' times.

- General code changes:
       - Update variable names to better reflect their functions! Refer to discussions below on how I optimized.

- Testing:
     - We make sure our tests cover the entire range of possibilities for optional and mandatory participants.
     - If time permits, we can extend test cases to test individual functions like sortEvents() or mergeEvents() to ensure each function does as intended

### CODE REVIEW

**Review by auzlee — CHANGES_REQUESTED** [review]
This looks much better! And thanks for the javadoc comments. I liked how you pulled the major components out into their own functions.

I had a few small nits/name suggestions, and then one more thought:
- Are we testing all the possible edge cases? It would be good to test more overlapping events, and sets where some events are "inside" other events.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,183 @@
...
+    public int compare(Event a, Event b) {
+      return Long.compare(a.getWhen().start(), b.getWhen().start());
+    }
+  };
+
+
+  /**
+   * Sorting events based on their start time in ascending order
+   */
+  public List<Event> sortEvents(Collection<Event> events, MeetingRequest request) {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: fair point

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,183 @@
...
+      }
+    }
+
+    Collections.sort(eventsWithPossibleConflicts, SORT_BY_START);
+    return eventsWithPossibleConflicts;
+  }
+
+
+  /**
+   * A comparator for sorting ranges by their start time in ascending order.
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: oops haha I will change to show it's a merge function

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,183 @@
...
+      return possibleTimesForMeeting;
+    } 
+    
+    if (request.getAttendees().isEmpty()) {
+      if (request.getOptionalAttendees().isEmpty()) {
+        Collection<TimeRange> possibleTimesForMeeting = Arrays.asList(TimeRange.WHOLE_DAY);
+        return possibleTimesForMeeting;
+      }
+
+      MeetingRequest newRequest = new MeetingRequest(request.getOptionalAttendees(), request.getDuration());
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:189]: :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,183 @@
...
+          unavailableTimeRanges.add(event1.getWhen());
+          unavailableTimeRanges.add(event2.getWhen());
+        }
+      } 
+    }
+
+    return unavailableTimeRanges;
+  }
+  
+  //convert blocked times to available times
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:83]: sure, I will make it private as there is no inherent need for it to be public

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,183 @@
...
+  public Collection<TimeRange> sectionTime(Collection<Event> events, MeetingRequest request, List<TimeRange> unavailableTimeRanges) {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:84]: I have renamed to findAvailableTimeRanges

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -121,6 +123,63 @@ public void everyAttendeeIsConsidered() {
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
+  public void everyAttendeeIsConsideredOptionalAllDay() {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:127]: changed to just attendeeIsConsideredOptionalAllDay

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -121,6 +123,63 @@ public void everyAttendeeIsConsidered() {
...
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void everyAttendeeIsConsideredOptional() {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:156]: changed to just attendeeIsConsideredOptional

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 8]
```diff
@@ -223,6 +282,35 @@ public void justEnoughRoom() {
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
+  public void justEnoughRoomOptional() {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:286]: changed to justEnoughRoomWithOptionalAttendee

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 9]
```diff
@@ -270,5 +358,55 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void noMandatoryAttendeesGaps() {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:363]: changed to noMandatoryAttendeesOptionalAttendeesWithGaps

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,183 @@
...
+  };
+
+
+  /**
+   * Sorting events based on their start time in ascending order
+   */
+  public List<Event> sortEvents(Collection<Event> events, MeetingRequest request) {
+    List<Event> eventsWithPossibleConflicts = new ArrayList<Event>();
+    for (Event event : events) {
+      if (request.getAttendees().containsAll(event.getAttendees())) {
```
[the remark this answers is not in the document]
Comment (gannerr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:44]: this is a great point. I will instead use !Collections.disjoint(request.getAttendees(), event.getAttendees()) to show if there exist ANY overlapping people.

**Review by auzlee — APPROVED** [review]
